sparse/coo/common.py

```diff
@@ -609,3 +609,72 @@ def nanreduce(x, method, identity=None, axis=None, keepdims=False, **kwargs):
     """
     arr = _replace_nan(x, method.identity if identity is None else identity)
     return arr.reduce(method, axis, keepdims, **kwargs)
+
+
+def roll(a, shift, axis=None):
+    """
+    Shifts elements of an array along specified axis. Elements that roll beyond
+    the last position are circulated and re-introduced at the first.
+
+    Parameters
+    ----------
+    x : COO
+        Input array
+    shift : int or tuple of ints
+        Number of index positions that elements are shifted. If a tuple is
+        provided, then axis must be a tuple of the same size, and each of the
+        given axes is shifted by the corresponding number. If an int while axis
+        is a tuple of ints, then broadcasting is used so the same shift is
+        applied to all axes.
+    axis : int or tuple of ints, optional
+        Axis or tuple specifying multiple axes. By default, the
+        array is flattened before shifting, after which the original shape is
+        restored.
+
+    Returns
+    -------
+    res : ndarray
+        Output array, with the same shape as a.
+    """
+    from .core import COO
+
+    # roll flattened array
+    if axis is None:
+        return roll(a.reshape((-1,)), shift, 0).reshape(a.shape)
+
+    # rolling does nothing for empty array
+    elif a.nnz == 0:
```

Comment:
Optional: Is this branch really required? I'd do the same for all special cases that aren't required. Will the above code already handle it?

Reply:
Ah yep let me get rid of that branch quickly